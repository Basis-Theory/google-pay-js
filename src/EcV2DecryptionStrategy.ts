import crypto from 'crypto';
import { parsePem } from 'eckey-utils';
import type { DecryptedMessage, SignedMessage } from './types';

export class EcV2DecryptionStrategy {
  // There is no type definitions for eckey-utils
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private readonly privateKey: any;

  public constructor(privateKeyPem: Buffer) {
    this.privateKey = parsePem(privateKeyPem.toString()).privateKey;
  }

  public decrypt(signedMessage: string): DecryptedMessage {
    const { ephemeralPublicKey, encryptedMessage, tag } = JSON.parse(
      signedMessage
    ) as SignedMessage;

    const sharedKey = this.getSharedKey(ephemeralPublicKey);
    const derivedKey = this.getDerivedKey(ephemeralPublicKey, sharedKey);

    const symmetricEncryptionKey = derivedKey.slice(0, 64);
    const macKey = derivedKey.slice(64);

    this.verifyMessageHmac(macKey, tag, encryptedMessage);

    const decryptedMessage = this.decryptMessage(
      encryptedMessage,
      symmetricEncryptionKey
    );

    return JSON.parse(decryptedMessage);
  }

  private getSharedKey(ephemeralPublicKey: string): string {
    /**
     *  Use the private key and the given ephemeralPublicKey to
     *  derive a 512-bit long shared key that uses ECIES-KEM.
     */
    const sharedKey = crypto.createECDH('prime256v1');

    sharedKey.setPrivateKey(this.privateKey);

    return sharedKey.computeSecret(ephemeralPublicKey, 'base64', 'hex');
  }

  private getDerivedKey(ephemeralPublicKey: string, sharedKey: string): string {
    return Buffer.from(
      crypto.hkdfSync(
        'sha256',
        Buffer.concat([
          Buffer.from(ephemeralPublicKey, 'base64'),
          Buffer.from(sharedKey, 'hex'),
        ]),
        Buffer.alloc(32),
        Buffer.from('Google', 'utf-8'),
        64
      )
    ).toString('hex');
  }

  private verifyMessageHmac(
    macKey: string,
    tag: string,
    encryptedMessage: string
  ): void {
    const hmac = crypto.createHmac('sha256', Buffer.from(macKey, 'hex'));

    hmac.update(Buffer.from(encryptedMessage, 'base64'));

    const calculatedTag = hmac.digest().toString('base64');

    if (calculatedTag !== tag) {
      throw new Error('Tag is not a valid MAC for the encrypted message');
    }
  }

  private decryptMessage(
    encryptedMessage: string,
    symmetricEncryptionKey: string
  ): string {
    const decipher = crypto.createCipheriv(
      'aes-256-ctr',
      Buffer.from(symmetricEncryptionKey, 'hex'),
      Buffer.alloc(16)
    );

    return Buffer.concat([
      decipher.update(Buffer.from(encryptedMessage, 'base64')),
      decipher.final(),
    ]).toString('utf-8');
  }
}
