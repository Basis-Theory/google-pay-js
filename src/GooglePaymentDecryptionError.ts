export class GooglePaymentDecryptionError extends Error {
  public constructor(message: string, public readonly errors?: Error[]) {
    super(message);
    this.name = 'GooglePaymentDecryptionError';
  }
}
