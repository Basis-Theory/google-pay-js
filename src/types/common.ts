interface PaymentMethodToken {
  protocolVersion: 'ECv2';
  signature: string;
  intermediateSigningKey: {
    signedKey: string;
    signatures: string[];
  };
  signedMessage: string;
}

interface SignedKey {
  keyValue: string;
  keyExpiration: string;
}

interface SignedMessage {
  encryptedMessage: string;
  ephemeralPublicKey: string;
  tag: string;
}

type PaymentMethodDetails = {
  pan: string;
  expirationMonth: number;
  expirationYear: number;
} & (
  | {
      authMethod: 'PAN_ONLY';
    }
  | {
      authMethod: 'CRYPTOGRAM_3DS';
      cryptogram: string;
      eciIndicator: string;
    }
);

interface DecryptedMessage {
  messageExpiration: string;
  messageId: string;
  paymentMethod: 'CARD';
  paymentMethodDetails: PaymentMethodDetails;
}

export type { PaymentMethodToken, DecryptedMessage, SignedKey, SignedMessage };
