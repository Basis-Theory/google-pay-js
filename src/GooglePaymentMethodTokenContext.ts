import { EcV2DecryptionStrategy } from './EcV2DecryptionStrategy';
import { GooglePaymentDecryptionError } from './GooglePaymentDecryptionError';
import type { DecryptedMessage, PaymentMethodToken } from './types';

interface GooglePayMerchantConfiguration {
  /**
   * (Optional) Merchant identifier, used for tracking errors.
   */
  identifier?: string;
  /**
   * Merchant private encryption key uploaded to Google Pay and Wallet Console
   */
  privateKeyPem: Buffer;
}
interface GooglePaymentMethodTokenContextOptions {
  merchants: GooglePayMerchantConfiguration[];
}

export class GooglePaymentMethodTokenContext {
  public constructor(
    private readonly options: GooglePaymentMethodTokenContextOptions
  ) {
    if (!options.merchants.length) {
      throw new GooglePaymentDecryptionError(
        'No merchant configuration provided for decryption context.'
      );
    }
  }

  public decrypt(token: PaymentMethodToken): DecryptedMessage {
    const errors = [];

    if (token.protocolVersion !== 'ECv2') {
      throw new Error(
        `Unsupported decryption for protocol version ${token.protocolVersion}`
      );
    }

    for (const merchant of this.options.merchants) {
      try {
        const strategy = new EcV2DecryptionStrategy(merchant.privateKeyPem);

        return strategy.decrypt(token.signedMessage);
      } catch (error) {
        error.merchantIdentifier = merchant.identifier;
        errors.push(error);
      }
    }

    throw new GooglePaymentDecryptionError(
      'Failed to decrypt payment data using provided merchant configuration(s).',
      errors
    );
  }
}
