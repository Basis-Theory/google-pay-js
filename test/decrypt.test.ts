import * as fs from 'fs';
import { GooglePaymentMethodTokenContext } from '../src';
import { PaymentMethodToken } from '../src/types';
import newToken from './fixtures/token.new.json';
import oldToken from './fixtures/token.old.json';

const newKey = fs.readFileSync('test/fixtures/key.new.pem');
const oldKey = fs.readFileSync('test/fixtures/key.old.pem');

describe('decrypt', () => {
  let newContext: GooglePaymentMethodTokenContext;
  let oldContext: GooglePaymentMethodTokenContext;
  let fullContext: GooglePaymentMethodTokenContext;

  beforeEach(() => {
    oldContext = new GooglePaymentMethodTokenContext({
      merchants: [
        {
          privateKeyPem: oldKey,
        },
      ],
    });
    newContext = new GooglePaymentMethodTokenContext({
      merchants: [
        {
          privateKeyPem: newKey,
        },
      ],
    });
    fullContext = new GooglePaymentMethodTokenContext({
      merchants: [
        {
          privateKeyPem: oldKey,
        },
        {
          privateKeyPem: newKey,
        },
      ],
    });
  });

  test('should throw when no merchant configuration is provided', () => {
    expect(
      () => new GooglePaymentMethodTokenContext({ merchants: [] })
    ).toThrow('No merchant configuration provided for decryption context.');
  });

  test('should throw when using old key', () => {
    expect(() => oldContext.decrypt(newToken as PaymentMethodToken)).toThrow(
      'Failed to decrypt payment data using provided merchant configuration(s).'
    );
  });

  test('should decrypt using new key', () => {
    expect(
      newContext.decrypt(newToken as PaymentMethodToken)
    ).toMatchSnapshot();
  });

  test('should support key rotation', () => {
    expect(
      fullContext.decrypt(oldToken as PaymentMethodToken)
    ).toMatchSnapshot();
    expect(
      fullContext.decrypt(newToken as PaymentMethodToken)
    ).toMatchSnapshot();
  });
});
