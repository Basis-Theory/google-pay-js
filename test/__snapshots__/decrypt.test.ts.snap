// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decrypt should decrypt using new key 1`] = `
{
  "messageExpiration": "1708953259025",
  "messageId": "AH2Ejtfux7LQy8R3HPjslN6NqKgbSp95fd4emwOs6qeGT78aJamj9WUvhMBRCRaLjuR_7Xqr7e1XTHfg0j5zePIrqG8RLrJu09AbEHg0GfLfJrImibYXWMvxTY0vTp5i5dTIoSSsxdG8",
  "paymentMethod": "CARD",
  "paymentMethodDetails": {
    "authMethod": "PAN_ONLY",
    "expirationMonth": 12,
    "expirationYear": 2026,
    "pan": "4111111111111111",
  },
}
`;

exports[`decrypt should support key rotation 1`] = `
{
  "messageExpiration": "1678476155759",
  "messageId": "AH2EjtcVAtpb_Nm4zwT6uJXPx8IHgKuxQEtVn7b8GJyxd_43iMIAarjXue1vjcaiJKMxIfzkp5JnD6gfdIB_q1rnPjR9ahV06-JvgGXya38xvgNQb4NSytKYoE3NxaPnMLt4vwI5iyMB",
  "paymentMethod": "CARD",
  "paymentMethodDetails": {
    "authMethod": "PAN_ONLY",
    "expirationMonth": 12,
    "expirationYear": 2028,
    "pan": "5555555555554444",
  },
}
`;

exports[`decrypt should support key rotation 2`] = `
{
  "messageExpiration": "1708953259025",
  "messageId": "AH2Ejtfux7LQy8R3HPjslN6NqKgbSp95fd4emwOs6qeGT78aJamj9WUvhMBRCRaLjuR_7Xqr7e1XTHfg0j5zePIrqG8RLrJu09AbEHg0GfLfJrImibYXWMvxTY0vTp5i5dTIoSSsxdG8",
  "paymentMethod": "CARD",
  "paymentMethodDetails": {
    "authMethod": "PAN_ONLY",
    "expirationMonth": 12,
    "expirationYear": 2026,
    "pan": "4111111111111111",
  },
}
`;
